crates/ty_python_semantic/src/types/diagnostic.rs

```diff
@@ -1650,7 +1651,24 @@ pub(super) fn report_unresolved_reference(context: &InferContext, expr_name_node
     };
 
     let ast::ExprName { id, .. } = expr_name_node;
-    builder.into_diagnostic(format_args!("Name `{id}` used when not defined"));
+    let mut diagnostic = builder.into_diagnostic(format_args!("Name `{id}` used when not defined"));
+    if let Some(version_added_to_builtins) = version_builtin_was_added(id) {
+        diagnostic.info(format_args!(
+            "`{id}` was added as a builtin in Python 3.{version_added_to_builtins}"
+        ));
+
+        // TODO: can we tell the user *why* we're inferring this target version?
+        // CLI flag? pyproject.toml? Python environment?
+        diagnostic.info(format_args!(
+            "The inferred target version of your project is Python {}",
+            Program::get(context.db()).python_version(context.db())
+        ));
```

Comment:
This shouldn't be too hard. We have the information if it comes from the CLI or configuration but we throw it away when converting it to a `ProgramSettings`.